### PR TITLE
Table query refactor

### PR DESCRIPTION
**Pull request by john-hill**

### CODE REVIEW

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SQLTranslatorUtils.java [hunk 1]
```diff
@@ -295,4 +317,261 @@ public static DerivedColumn createDerivedColumn(MysqlFunction mysqlFunction) {
...
+				BooleanPrimary newPrimary = new TableQueryParser(builder.toString()).booleanPrimary();
+				booleanPrimary.replaceSearchCondition(newPrimary.getSearchCondition());
+			} catch (ParseException e) {
+				throw new IllegalArgumentException(e);
+			}
+		}
+	}
+
+	/**
+	 * Translate a HasReferencedColumn for the select clause.
```
Comment (kimyen) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SQLTranslatorUtils.java:525]: Translate from what to what?

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SQLUtils.java [hunk 2]
```diff
@@ -611,28 +611,41 @@ private static void appendColumnName(String prefix, String subName, ColumnModel
...
 			}
 			builder.append(tableName).append(".");
-			appendColumnName(subName, column, builder);
+			appendColumnName(subName, columnId, builder);
 		}
 	}
+	
+	/**
+	 * Create a double clause.
+	 * @param columnId
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SQLUtils.java:639]: sp

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SqlQuery.java [hunk 3]
```diff
@@ -125,19 +133,23 @@ public void init(QuerySpecification parsedModel, List<ColumnModel> tableSchema,
...
+			transformedModel.replaceSelectList(expandedSelectList);
 			this.includesRowIdAndVersion = true;
 		}else{
 			this.includesRowIdAndVersion = false;
 		}
-
-		this.outputSQL = SQLTranslatorUtils.translate(transformedModel, this.parameters, this.columnNameToModelMap);
-
+//		this.outputSQL = SQLTranslatorUtils.translate(transformedModel, this.parameters, this.columnNameToModelMap);
+		SQLTranslatorUtils.translateModel(transformedModel, parameters, columnNameToModelMap);
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SqlQuery.java]: remove comment

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SqlQuery.java [hunk 4]
```diff
@@ -125,19 +133,23 @@ public void init(QuerySpecification parsedModel, List<ColumnModel> tableSchema,
...
+			SelectList expandedSelectList = SQLTranslatorUtils.addRowIdAndVersionToSelect(this.transformedModel.getSelectList());
+			transformedModel.replaceSelectList(expandedSelectList);
 			this.includesRowIdAndVersion = true;
 		}else{
 			this.includesRowIdAndVersion = false;
 		}
-
-		this.outputSQL = SQLTranslatorUtils.translate(transformedModel, this.parameters, this.columnNameToModelMap);
-
+//		this.outputSQL = SQLTranslatorUtils.translate(transformedModel, this.parameters, this.columnNameToModelMap);
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SqlQuery.java:150]: Remove comment.

File: lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/SQLQueryTest.java [hunk 5]
```diff
@@ -127,7 +128,7 @@ public void testTranslateColumnReferenceRowVersionIgnoreCase() throws ParseExcep
 		assertEquals("ROW_VERSION", visitor.getSql());
 	}
 	
-	@Test
+	@Test (expected=IllegalArgumentException.class)
 	public void testSelectStarEmtpySchema() throws ParseException{
 		tableSchema = new LinkedList<ColumnModel>();
 		SqlQuery translator = new SqlQuery("select * from syn123", tableSchema);
```
Comment (john-hill) [inline lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/SQLQueryTest.java:134]: remove last line

File: lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/SQLQueryTest.java [hunk 6]
```diff
@@ -651,4 +664,95 @@ public void testTypeSetFunctionDoubles() throws Exception {
 		assertEquals(TableModelUtils.createSelectColumn("COUNT(doubletype)", ColumnType.INTEGER, null), translator.getSelectColumns()
 				.get(4));
 	}
+		
+	@Test
+	public void testPLFM_3864() throws Exception{
```
Comment (john-hill) [inline lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/SQLQueryTest.java:669]: add a link to the jira of the javadoc

File: lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/SQLQueryTest.java [hunk 7]
```diff
@@ -651,4 +664,95 @@ public void testTypeSetFunctionDoubles() throws Exception {
...
+	
+	@Test
+	public void testPLFM_3867() throws ParseException{
+		SqlQuery translator = new SqlQuery("select foo from syn123 where foo = \"a\"", tableSchema);
+		assertEquals("SELECT _C111_, ROW_ID, ROW_VERSION FROM T123 WHERE _C111_ = :b0", translator.getOutputSQL());
+		assertEquals("a", translator.getParameters().get("b0"));
+	}
+	
+	@Test
+	public void testTranslateNotNaN() throws ParseException{
```
Comment (john-hill) [inline lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/SQLQueryTest.java:723]: Translate not isNaN()

File: lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/model/ValueExpressionPrimary.java [hunk 8]
```diff
@@ -65,6 +65,28 @@ public void toSql(StringBuilder builder) {
...
+				return setFunctionSpecification.getValueExpression().getFirstElementOfType(HasQuoteValue.class);
+			}
+		}else{
+			// This is not a function so get the first unquoted.
+			return this.getFirstElementOfType(HasQuoteValue.class);
+		}
+	}
+
+	@Override
+	public boolean isReferenceInFunction() {
```
Comment (john-hill) [inline lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/model/ValueExpressionPrimary.java:86]: throw an exception if reference is null for this method

**Pull request merged**